<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Cytoscape Tutorial: R driving CyREST</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.0.13/css/fa-svg-with-js.css" rel="stylesheet" />
<script src="site_libs/font-awesome-5.0.13/js/fontawesome-all.min.js"></script>
<script src="site_libs/font-awesome-5.0.13/js/fa-v4-shims.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Cytoscape Rmd Notebooks</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="http://cytoscape.github.io/cytoscape-automation/for-scripters/R/notebooks">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-list"></span>
     
    Table of Contents
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Tutorial Notebooks</li>
    <li>
      <a href="basic-fundamentals.html">Basic fundamentals</a>
    </li>
    <li>
      <a href="http://bioconductor.org/packages/release/bioc/vignettes/RCy3/inst/doc/Overview-of-RCy3.html">Overview of RCy3</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Analysis Notebooks</li>
    <li>
      <a href="advanced-cancer-networks-and-data-rcy3.html">Advanced cancer networks and data</a>
    </li>
    <li>
      <a href="https://github.com/CyTargetLinker/cytargetlinker-tutorials/blob/master/R/UseCase2/Usecase2_DEGs_Pathway_Network.Rmd">CyTargetLinker: use case 2, gene-pathway extension</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Cytoscape Tutorial: R driving CyREST</h1>

</div>


<p>This is an <a href="http://rmarkdown.rstudio.com">R Markdown</a> Notebook that encapsulates a common bioinformatics workflow.</p>
<p>We will retrieve disease networks from a public database and apply gene expression and tumor mutation datasets for network analysis and visualization. Network files and images will be generated in multiple formats for sharing and publishing.</p>
<div id="basic-steps-to-this-workflow." class="section level3">
<h3>Basic steps to this workflow.</h3>
<ul>
<li>Install requisite libraries and connect to Cytoscape.</li>
<li>Query the STRING database for network associated with the diseases</li>
<li>Explore interactivity using CyREST, Commands and RCy3</li>
<li>Load data into Cytoscape and visualize data on networks</li>
<li>Select subnetworks based on data</li>
<li>Save, share and publish network results</li>
</ul>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/CytoscapeAutomation_3.png" alt="Cytoscape automation resource diagram" />
<p class="caption">Cytoscape automation resource diagram</p>
</div>
</div>
<div id="tested-with-the-following-versions" class="section level3">
<h3>Tested with the following versions:</h3>
<ul>
<li>R 3.3.3</li>
<li>RStudio 1.0.143</li>
<li>Cytoscape 3.6.0</li>
<li>CyREST 3.6.0</li>
<li>stringApp 1.2.0</li>
<li>Operating Systems tested:
<ul>
<li>Mac 10.10.5</li>
<li>Windows 10</li>
<li>Linux (may need to install R-devel)</li>
</ul></li>
</ul>
<div id="section" class="section level48">
<p></p>
</div>
</div>
<div id="step-0-are-you-ready" class="section level1">
<h1>Step 0: Are You Ready?</h1>
<div id="section-1" class="section level48">
<p></p>
<p>This section will load the appropriate libraries and ensure they are functional.</p>
</div>
<div id="install-libs-and-utils" class="section level3">
<h3>Install libs and utils</h3>
<p>Please run check-library-installation.R for libs and tests:</p>
<p><a href="https://github.com/cytoscape/cytoscape-automation/blob/master/for-scripters/R/check-library-installation.R" class="uri">https://github.com/cytoscape/cytoscape-automation/blob/master/for-scripters/R/check-library-installation.R</a></p>
<p><strong>If you are in RStudio</strong> try executing this chunk by clicking the <em>Run</em> button within the chunk or by placing your cursor inside it and pressing <em>Cmd+Shift+Enter</em>.</p>
<pre class="r"><code>library(RColorBrewer)</code></pre>
</div>
<div id="setup-cytoscape" class="section level3">
<h3>Setup Cytoscape</h3>
<ul>
<li>Launch Cytoscape on your local machine. If you haven’t already installed Cytoscape, then download the latest version from <a href="http://cytoscape.org" class="uri">http://cytoscape.org</a>.</li>
<li>Leave Cytoscape running in the background during the remainder of the tutorial.</li>
<li>Install the STRING app, if you haven’t already: <a href="http://apps.cytoscape.org/apps/stringapp" class="uri">http://apps.cytoscape.org/apps/stringapp</a></li>
</ul>
<pre class="r"><code>installApp(&#39;STRING&#39;)</code></pre>
<p>Please run check-cytoscape-connection.R for tests and initial demo:</p>
<p><a href="https://github.com/cytoscape/cytoscape-automation/blob/master/for-scripters/R/check-cytoscape-connection.R" class="uri">https://github.com/cytoscape/cytoscape-automation/blob/master/for-scripters/R/check-cytoscape-connection.R</a></p>
<div id="test-connection-to-cytoscape-and-installed-apps" class="section level4">
<h4>Test connection to Cytoscape and installed apps</h4>
<p><strong>port</strong> number needs to match value of Cytoscape property: rest.port (see Edit&gt;Preferences&gt;Properties…). Default is 1234.</p>
<pre class="r"><code>cytoscapeVersionInfo()</code></pre>
<div id="section-2" class="section level48">
<p></p>
</div>
</div>
</div>
</div>
<div id="step-1-getting-disease-networks" class="section level1">
<h1>Step 1: Getting Disease Networks</h1>
<div id="section-3" class="section level48">
<p></p>
<p>Use Cytoscape to query the STRING database for networks of genes associated with breast cancer and ovarian cancer.</p>
<p><strong>If the STRING app is not installed, no error is reported, but your network will be empty</strong></p>
</div>
<div id="query-string-database-by-disease-to-generate-networks" class="section level3">
<h3>Query STRING database by disease to generate networks</h3>
<div id="breast-cancer" class="section level4">
<h4>Breast cancer</h4>
<pre class="r"><code>string.cmd = &#39;string disease query disease=&quot;breast cancer&quot; cutoff=0.9 species=&quot;Homo sapiens&quot; limit=150&#39;
commandsGET(string.cmd)</code></pre>
<p>Here we are using Cytoscape’s command line syntax, converting it to a CyREST url, and then making a GET request.</p>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/brc.png" style="height:100.0%" />

</div>
</div>
<div id="ovarian-cancer" class="section level4">
<h4>Ovarian cancer</h4>
<pre class="r"><code>string.cmd = &#39;string disease query disease=&quot;ovarian cancer&quot; cutoff=0.9 species=&quot;Homo sapiens&quot; limit=150&#39;
commandsGET(string.cmd)</code></pre>
<p>This time, we skipped a step and used the helper function <em>commandRun</em> to make the GET request from the command string.</p>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/ovc.png" style="height:100.0%" />

</div>
<div id="section-4" class="section level48">
<p></p>
</div>
</div>
</div>
</div>
<div id="step-2-interacting-with-cytoscape" class="section level1">
<h1>Step 2: Interacting with Cytoscape</h1>
<div id="section-5" class="section level48">
<p></p>
<p>Now that we’ve got a couple networks into Cytoscape, let’s see what we can go with them from R…</p>
</div>
<div id="get-list-of-networks" class="section level3">
<h3>Get list of networks</h3>
<pre class="r"><code>getNetworkList()</code></pre>
<div id="challenge" class="section level4">
<h4>CHALLENGE</h4>
<p>If the <strong>listNetworks()</strong> helper function didn’t exist, how could you get the list of networks in R? <em>Hint:</em> the command is <strong>network list</strong>.</p>
</div>
</div>
<div id="layout-network" class="section level3">
<h3>Layout network</h3>
<pre class="r"><code>layoutNetwork(layout.name=&#39;circular&#39;) </code></pre>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/ovc2.png" style="height:100.0%" />

</div>
<div id="list-of-layout-algorithms-available" class="section level4">
<h4>List of layout algorithms available</h4>
<pre class="r"><code>getLayoutNames()</code></pre>
</div>
<div id="challenge-1" class="section level4">
<h4>CHALLENGE</h4>
<p>Pick a different layout algorithm and execute <strong>applyLayout()</strong>.</p>
</div>
<div id="layout-with-parameters" class="section level4">
<h4>Layout with parameters!</h4>
<pre class="r"><code>getLayoutPropertyNames(layout.name=&#39;force-directed&#39;)
layoutNetwork(&#39;force-directed defaultSpringCoefficient=0.0000008 defaultSpringLength=70&#39;)</code></pre>
<p>These <em>commandHelp</em> calls will allow you to interrogate the layouts and parameters available in your active Cytoscape session. This includes any layouts added by apps you’ve installed!</p>
</div>
</div>
<div id="get-table-data-from-network" class="section level3">
<h3>Get table data from network</h3>
<p>Now, let’s look at the tablular data associated with our STRING networks…</p>
<pre class="r"><code>getTableColumnNames(&#39;node&#39;)</code></pre>
<p>One of the great things about the STRING database is all the node and edge attriubtes they provide. Let’s pull some of it into R to play with…</p>
<div id="retrieve-disease-scores" class="section level4">
<h4>Retrieve disease scores</h4>
<p>We can retrieve any set of columns from Cytoscape and store them as an R data frame.</p>
<pre class="r"><code>disease.score.table &lt;- getTableColumns(&#39;node&#39;,&#39;disease score&#39;)
disease.score.table</code></pre>
</div>
<div id="plot-distribution-and-pick-threshold" class="section level4">
<h4>Plot distribution and pick threshold</h4>
<p>Now you can use R like you normally would explore the data.</p>
<pre class="r"><code>plot(factor(row.names(disease.score.table)),disease.score.table[,1], ylab=colnames(disease.score.table)[1])
summary(disease.score.table)</code></pre>
</div>
</div>
<div id="generate-subnetworks" class="section level3">
<h3>Generate subnetworks</h3>
<p>In order to reflect your exploration back onto the network, let’s generate subnetworks…</p>
<p>…from top quartile of ‘disease score’</p>
<pre class="r"><code>top.quart &lt;- quantile(disease.score.table[,1], 0.75)
top.nodes &lt;- row.names(disease.score.table)[which(disease.score.table[,1]&gt;top.quart)]
createSubnetwork(top.nodes,subnetwork.name =&#39;top disease quartile&#39;)
#returns a Cytoscape network SUID</code></pre>
<p>…of connected nodes only</p>
<pre class="r"><code>createSubnetwork(edges=&#39;all&#39;,subnetwork.name=&#39;top disease quartile connected&#39;)  #handy way to exclude unconnected nodes!</code></pre>
<p>…from first neighbors of top 3 genes, using the network connectivity together with the data to direct discovery.</p>
<pre class="r"><code>setCurrentNetwork(network=&quot;String Network - ovarian cancer&quot;)
top.nodes &lt;- row.names(disease.score.table)[tail(order(disease.score.table[,1]),3)]
selectNodes(nodes=top.nodes)
selectFirstNeighbors()
createSubnetwork(&#39;selected&#39;, subnetwork.name=&#39;top disease neighbors&#39;) # selected nodes, all connecting edges (default)</code></pre>
<p>…from diffusion algorithm starting with top 3 genes, using the network connectivity in a more subtle way than just first-degree neighbors.</p>
<pre class="r"><code>setCurrentNetwork(network=&quot;String Network - ovarian cancer&quot;)
selectNodes(nodes=top.nodes)
commandsGET(&#39;diffusion diffuse&#39;) # diffusion!
createSubnetwork(&#39;selected&#39;,subnetwork.name = &#39;top disease diffusion&#39;)
layoutNetwork(&#39;force-directed&#39;)</code></pre>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/ovc4.png" style="height:100.0%" />

</div>
<div id="challenge-2" class="section level4">
<h4>CHALLENGE</h4>
<p>Okay, now for a real challenge… pull in a new column of data from Cytoscape (provided by STRING) and use it to make a new selection. Optionally, expand on that selection. Then create a new subnetwork and rename it. *Pro-<a href="tip:*" class="uri">tip:*</a> don’t forget to <strong>setCurrentNetwork()</strong> the correct parent network before getting table column data and making selections.</p>
<div id="section-6" class="section level48">
<p></p>
</div>
</div>
</div>
</div>
<div id="step-3-visualizing-data-on-networks" class="section level1">
<h1>Step 3: Visualizing data on networks</h1>
<div id="section-7" class="section level48">
<p></p>
</div>
<div id="load-datasets" class="section level3">
<h3>Load datasets</h3>
<p>Downloaded TCGA data from <a href="https://portal.gdc.cancer.gov/" class="uri">https://portal.gdc.cancer.gov/</a> and preprocessed as R objects. Also available via each TCGA publication, e.g.:</p>
<ul>
<li>Breast: <a href="https://tcga-data.nci.nih.gov/docs/publications/brca_2012/" class="uri">https://tcga-data.nci.nih.gov/docs/publications/brca_2012/</a></li>
<li>Ovarian: <a href="https://tcga-data.nci.nih.gov/docs/publications/ov_2011/" class="uri">https://tcga-data.nci.nih.gov/docs/publications/ov_2011/</a></li>
</ul>
<pre class="r"><code>load(&quot;data/tutorial-ovc-expr-mean-dataset.robj&quot;)
load(&quot;data/tutorial-ovc-mut-dataset.robj&quot;)
load(&quot;data/tutorial-brc-expr-mean-dataset.robj&quot;)
load(&quot;data/tutorial-brc-mut-dataset.robj&quot;)</code></pre>
</div>
<div id="breast-cancer-datset" class="section level3">
<h3>Breast Cancer Datset</h3>
<p>These datasets are similar to the data frames you normarlly encounter in R. For diversity, one using row.names to store corresponding gene names and the other uses the first column. Both are easy to import into Cytoscape.</p>
<pre class="r"><code>str(brc.expr)  # gene names in row.names of data.frame
str(brc.mut)  # gene names in column named &#39;Hugo_Symbol&#39;</code></pre>
<p>Let’s return to the Breast Cancer network…</p>
<pre class="r"><code>setCurrentNetwork(network=&quot;String Network - breast cancer&quot;)
layoutNetwork(&#39;force-directed&#39;) #uses same settings as previously set</code></pre>
<p>…and use the helper function from RCy3 called <em>loadTableData</em></p>
<pre class="r"><code>?loadTableData
loadTableData(brc.expr,table.key.column = &quot;display name&quot;)  #default data.frame key is row.names
loadTableData(brc.mut,&#39;Hugo_Symbol&#39;,table.key.column = &quot;display name&quot;)  #specify column name if not default</code></pre>
<div id="visual-styles" class="section level4">
<h4>Visual styles</h4>
<p>Let’s create a new style to visualize our imported data …starting with the basics, we will specify a few defaults and obvious mappings in a custom style all our own.</p>
<pre class="r"><code>style.name = &quot;dataStyle&quot;
defaults.list &lt;- list(NODE_SHAPE=&quot;ellipse&quot;,
                 NODE_SIZE=60,
                 NODE_FILL_COLOR=&quot;#AAAAAA&quot;,
                 EDGE_TRANSPARENCY=120)
node.label.map &lt;- mapVisualProperty(&#39;node label&#39;,&#39;display name&#39;,&#39;p&#39;) # p for passthrough; nothing else needed
createVisualStyle(style.name, defaults.list, list(node.label.map))
setVisualStyle(style.name=style.name)</code></pre>
</div>
<div id="challenge-3" class="section level4">
<h4>CHALLENGE</h4>
<pre class="r"><code>?createVisualStyle</code></pre>
<p>What is the difference between <em>defaults</em> and <em>mappings</em>? Why was node label set by passthrough? What would happen if you set <strong>NODE_LABEL=“name”</strong> as a default? Where can you find the list of all possible Visual Properties?</p>
<div id="visualize-expression-data" class="section level5">
<h5>Visualize expression data</h5>
<p>Now let’s update the style with a mapping for mean expression. The first step is to grab the column data from Cytoscape and pull out the min and max to define our data mapping range of values.</p>
<pre class="r"><code>brc.expr.network = getTableColumns(&#39;node&#39;,&#39;expr.mean&#39;)  
min.brc.expr = min(brc.expr.network[,1],na.rm=TRUE)
max.brc.expr = max(brc.expr.network[,1],na.rm=TRUE)
data.values = c(min.brc.expr,0,max.brc.expr)</code></pre>
<p>Next, we use the RColorBrewer package to help us pick good colors to pair with our data values.</p>
<pre class="r"><code>display.brewer.all(length(data.values), colorblindFriendly=TRUE, type=&quot;div&quot;) # div,qual,seq,all
node.colors &lt;- c(rev(brewer.pal(length(data.values), &quot;RdBu&quot;)))</code></pre>
<p>Finally, we use the handy <em>mapVisualProperty</em> function to construct the data object that CyREST needs to specify style mappings and then we’ll send them off to Cytoscape with <em>updateStyleMappings</em>.</p>
<pre class="r"><code>setNodeColorMapping(&#39;expr.mean&#39;, data.values, node.colors, style.name=style.name)</code></pre>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/brc2.png" style="height:100.0%" />

</div>
<p><strong>Pro-tip: depending on your data, it may be better to balance your color range over negative and positive values bounded by the largest min or max data value, so that color intensity scales similarly in both directions.</strong></p>
</div>
</div>
<div id="challenge-4" class="section level4">
<h4>CHALLENGE</h4>
<p>Update the style with a balanced data visualization range based on the largest min or max value. <em>Hints:</em> redefine data.values, redefine node.fill.map, rerun updateStyleMapping().</p>
</div>
<div id="visualize-mutation-data" class="section level4">
<h4>Visualize mutation data</h4>
<p>OK, now let’s update with a mapping for mutation. Here are all the same steps, but this time mapping mutation counts to <em>both</em> node border width and color.</p>
<pre class="r"><code>brc.mut.network = getTableColumns(&#39;node&#39;,&#39;mut_count&#39;)
min.brc.mut = min(brc.mut.network[,1],na.rm=TRUE)
max.brc.mut = max(brc.mut.network[,1],na.rm=TRUE)
data.values = c(min.brc.mut,20,max.brc.mut)
display.brewer.all(length(data.values), colorblindFriendly=TRUE, type=&quot;seq&quot;) 
border.colors &lt;- c(brewer.pal(3, &quot;Reds&quot;))
setNodeBorderColorMapping(&#39;mut_count&#39;,data.values,border.colors,style.name=style.name)
border.width &lt;- c(2,4,8)
setNodeBorderWidthMapping(&#39;mut_count&#39;,data.values,border.width,style.name=style.name)</code></pre>
<p>This is a useful pair of visual properties to map to a single data column. See why?</p>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/brc3.png" style="height:100.0%" />

</div>
</div>
<div id="subnetwork-based-on-diffusion-from-heavily-mutated-nodes" class="section level4">
<h4>Subnetwork based on diffusion from heavily mutated nodes</h4>
<p>Now, let’s pull in what we learned about subnetwork selection and apply it here…</p>
<pre class="r"><code>top.mut &lt;- (brc.mut$Hugo_Symbol)[tail(order(brc.mut$mut_count),2)]
top.mut
selectNodes(nodes=top.mut,&#39;display name&#39;)
commandsGET(&#39;diffusion diffuse&#39;) 
createSubnetwork(&#39;selected&#39;,subnetwork.name = &#39;top mutated diffusion&#39;)
layoutNetwork(&#39;force-directed&#39;)</code></pre>
<p>The top mutated genes are based on TCGA data and the diffusion algorithm is operating based on the network connectivity from STRING data, leading to a focused subnetwork view of critical Breast Cancer genes with mean patient expression data mapped to fill color. Now <em>that’s</em> data integration!</p>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/brc4.png" style="height:100.0%" />

</div>
<p><strong>Pro-tip: You can generate a legend for this in Cytoscape Style tab &gt; Options &gt; Create style… This is no yet available as a command. Coming soon!</strong></p>
</div>
</div>
<div id="ovarian-cancer-datset" class="section level3">
<h3>Ovarian Cancer Datset</h3>
<p>But what about the other network and datasets? Do we have to repeat <em>all</em> of those steps again? Actually, no!</p>
<p>First, let’s switch back over to the Ovarian Cancer network and load our data.</p>
<pre class="r"><code>setCurrentNetwork(network=&quot;String Network - ovarian cancer&quot;)
clearSelection()
str(ovc.expr)  # gene names in row.names of data.frame
str(ovc.mut)  # gene names in column named &#39;Hugo_Symbol&#39;

loadTableData(ovc.expr, table.key.column = &#39;display name&#39;)
loadTableData(ovc.mut,&#39;Hugo_Symbol&#39;, table.key.column = &#39;display name&#39;)</code></pre>
<p><strong>Because we used the same column names in our original data frames, now we can simply apply the <em>same</em> visual style created above!</strong></p>
<pre class="r"><code>setVisualStyle(style.name=style.name)</code></pre>
<div class="figure">
<img src="https://cytoscape.github.io/cytoscape-automation/for-scripters/R/data/img/ovc3.png" style="height:100.0%" />

</div>
<p>Reusing the same style for both breast and ovarian cancers, we can compare the relative expression and mutation counts across the two datasets. For example, notice in the case of ovarian cancer: <strong>decreased</strong> range of mean expression and <strong>fewer</strong> mega-mutated genes.</p>
<div id="section-8" class="section level42">
<p></p>
</div>
</div>
</div>
<div id="step-4-saving-sharing-and-publishing" class="section level1">
<h1>Step 4: Saving, sharing and publishing</h1>
<div id="section-9" class="section level42">
<p></p>
</div>
<div id="saving-a-cytoscape-session-file" class="section level3">
<h3>Saving a Cytoscape session file</h3>
<p>Session files save <em>everything</em>. As with most project software, we recommend saving often!</p>
<pre class="r"><code>saveSession(&#39;tutorial_session&#39;) #.cys</code></pre>
<p><strong>Note: If you don’t specify a complete path, the files will be save relative to your Cytoscape installation directory, e.g., /Applications/Cytoscape_v3.6.0/… or somewhere you don’t have write permissions. So, it’s probably better to specify one…</strong></p>
<pre class="r"><code>full.path=paste(getwd(),&#39;tutorial_session&#39;,sep=&#39;/&#39;)
saveSession(filename=full.path) #.cys</code></pre>
</div>
<div id="saving-high-resolution-image-files" class="section level3">
<h3>Saving high resolution image files</h3>
<p>You can export extremely high resolution images, including vector graphic formats.</p>
<pre class="r"><code>full.path=paste(getwd(),&#39;tutorial_image2&#39;,sep=&#39;/&#39;)
exportImage(filename=full.path, type = &#39;PDF&#39;) #.pdf
?exportImage</code></pre>
</div>
<div id="exporting-interactive-web-pages-via-cytoscape.js" class="section level3">
<h3>Exporting interactive web pages via cytoscape.js</h3>
<p>You can export the network model in multiple standard formats. And if you want to get really fancy, you can export the model and style in cytoscape.js formats and host interactive views of your network analysis and visualization results online!</p>
<pre class="r"><code>full.path=paste(getwd(),&#39;tutorial_network&#39;,sep=&#39;/&#39;)
exportNetwork(filename=full.path, type=&#39;CYJS&#39;)  #.cyjs
?exportNetwork

full.path=paste(getwd(),&#39;tutorial_style&#39;,sep=&#39;/&#39;)
exportVisualStyles(filename=full.path, type=&#39;json&#39;) #.json; current style
?exportVisualStyles</code></pre>
<p>More details in Cytoscape manual chapter on <a href="http://manual.cytoscape.org/en/stable/Export_Your_Data.html#as-an-interactive-web-application-new-in-3-2-0">Interactive Web Applications</a></p>
<p>And in Cytoscape, you can File&gt;Export as Web Page… to produce a folder of cytoscape.js files and html.</p>
</div>
<div id="additional-network-analysis-and-visualization-options" class="section level3">
<h3>Additional network analysis and visualization options</h3>
<ul>
<li>Mapping edge attributes to edge thickness, color and/or transparency (e.g., score or experiments)
<ul>
<li>EDGE_WIDTH, EDGE_COLOR, EDGE TRANSPARENCY</li>
</ul></li>
<li>Edge bundling
<ul>
<li>bundleEdges()</li>
</ul></li>
<li>Directional edge animation
<ul>
<li>applyStyle(‘Marquee’)</li>
<li>updateStyleDefaults(style.name,list(‘Edge Line Type’=‘Marquee_Dash’))</li>
</ul></li>
<li>Visualize two or more datasets as split-fill or pie chart, or patient-specific data as radial Circos plot
<ul>
<li>pending support for charts…</li>
</ul></li>
<li>Merge two (or more) networks
<ul>
<li>copyCytoscapeNetwork() or ‘network clone’; then addGraphToGraph() (Union only)</li>
<li>using igraph: intersection(g1,g2), union(), difference(); followed by toCytoscape() and send2cy()</li>
</ul></li>
</ul>
<div id="resources-for-all-available-functions-and-commands" class="section level4">
<h4>Resources for all available functions and commands</h4>
<pre class="r"><code>help(package=RCy3) # RCy3 helper functions
cyrestAPI()  # CyREST API
commandsAPI()  # CyREST Commands API</code></pre>
<div id="section-10" class="section level42">
<p></p>
</div>
</div>
</div>
</div>
<div id="step-5-wrapping-up" class="section level1">
<h1>Step 5: Wrapping up</h1>
<div id="section-11" class="section level42">
<p></p>
<p>You should now know how to…</p>
<ul>
<li>Connect to Cytoscape from R</li>
<li>Retrieve disease networks from STRING</li>
<li>Interact with Cytoscape via commands, CyREST, and RCy3</li>
<li>Load various data types onto networks</li>
<li>Set visual style defaults and mappings</li>
<li>Perform selection and create subnetworks</li>
<li>Save, share and publish your networks</li>
<li>Find more information on Cytoscape automation</li>
</ul>
</div>
<div id="resources" class="section level4">
<h4>Resources</h4>
<ul>
<li>Growing collection of <a href="https://github.com/cytoscape/cytoscape-automation/">cytoscape automation resources</a>, including <strong>this</strong> tutorial</li>
<li>Growing collection of <a href="https://github.com/cytoscape/cytoscape-tutorials">cytoscape tutorial materials</a></li>
<li>Active community on <a href="https://groups.google.com/forum/#!forum/cytoscape-helpdesk">cytoscape-helpdesk</a> for any Cytoscape related questions and suggestions</li>
</ul>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
